File: src/main/java/io/github/svegon/utils/fast/util/chars/booleans/CharCompressedTableSet.java
package io.github.svegon.utils.fast.util.chars.booleans;

import io.github.svegon.utils.ConditionUtil;
import io.github.svegon.utils.collections.ArrayUtil;
import io.github.svegon.utils.fast.util.chars.ImprovedCharCollection;
import io.github.svegon.utils.fast.util.chars.OnNextComputeCharIterator;
import io.github.svegon.utils.math.MathUtil;
import it.unimi.dsi.fastutil.bytes.ByteArrays;
import it.unimi.dsi.fastutil.bytes.ByteListIterator;
import it.unimi.dsi.fastutil.chars.AbstractCharSet;
import it.unimi.dsi.fastutil.chars.CharCollection;
import it.unimi.dsi.fastutil.chars.CharIterable;
import it.unimi.dsi.fastutil.chars.CharIterator;
import net.jcip.annotations.NotThreadSafe;
import org.jetbrains.annotations.NotNull;

@NotThreadSafe
public class CharCompressedTableSet extends AbstractCharSet implements ImprovedCharCollection {
    private byte[] table;

    public CharCompressedTableSet() {
        table = ByteArrays.DEFAULT_EMPTY_ARRAY;
    }

    public CharCompressedTableSet(char expectedMaxKey) {
        table = expectedMaxKey == 0 ? ByteArrays.EMPTY_ARRAY : new byte[(expectedMaxKey >>> 3)
                + (expectedMaxKey & 7) != 0 ? 1 : 0];
    }

    public CharCompressedTableSet(final @NotNull CharIterable elements) {
        this(elements.iterator());
    }

    public CharCompressedTableSet(final @NotNull CharIterator elements) {
        this((char) 16);

        while (elements.hasNext()) {
            add(elements.nextChar());
        }
    }

    @Override
    public CharIterator iterator() {
        return new OnNextComputeCharIterator() {
            final ByteListIterator it = ArrayUtil.asList(table).iterator();

            @Override
            protected char computeNext() {
                byte b;

                while (it.hasNext()) {
                    if ((b = it.nextByte()) != 0) {
                        return (char) ((it.nextIndex() << 3) | MathUtil.highestOneBit(b));
                    }
                }

                finish();
                return 0;
            }
        };
    }

    @Override
    public int size() {
        return ArrayUtil.stream(table).mapToInt(Integer::bitCount).sum();
    }

    @Override
    public boolean add(char k) {
        int index = k >>> 3;
        int mod = k & 7;
        table = ByteArrays.ensureCapacity(table, index + 1);
        return table[k] != (table[index] |= 1 << mod);
    }

    @Override
    public boolean contains(char k) {
        int index = k >>> 3;

        if (index >= table.length) {
            return false;
        }

        return ConditionUtil.hasFlag(table[index], k & 7);
    }

    @Override
    public boolean remove(char k) {
        int index = k >>> 3;

        if (index >= table.length) {
            return false;
        }

        return table[index] != (table[index] &= ~(1 << (k & 7)));
    }

    @Override
    public boolean removeAll(final @NotNull CharCollection c) {
        boolean modified = false;

        for (char s : c) {
            modified |= remove(s);
        }

        return modified;
    }

    @Override
    public boolean retainAll(final @NotNull CharCollection c) {
        return removeIf((e) -> !c.contains(e));
    }
}

File: src/main/kotlin/io/github/svegon/mclientapi/util/collections/chars/CharCompressedTableSet.kt
package io.github.svegon.mclientapi.util.collections.chars

import io.github.svegon.utils.ConditionUtil
import io.github.svegon.utils.collections.ArrayUtil
import io.github.svegon.utils.fast.util.chars.OnNextComputeCharIterator
import io.github.svegon.utils.math.MathUtil
import it.unimi.dsi.fastutil.bytes.ByteArrays
import it.unimi.dsi.fastutil.bytes.ByteListIterator
import it.unimi.dsi.fastutil.chars.*
import it.unimi.dsi.fastutil.chars.CharIterator
import oshi.annotation.concurrent.NotThreadSafe

@NotThreadSafe
class CharCompressedTableSet : AbstractCharSet {
    private var table: ByteArray

    constructor() {
        table = ByteArrays.DEFAULT_EMPTY_ARRAY
    }

    constructor(expectedMaxKey: Char) {
        table = if (expectedMaxKey.code == 0) ByteArrays.EMPTY_ARRAY else ByteArray(
            if ((expectedMaxKey.code ushr 3)
                + (expectedMaxKey.code and 7) != 0
            ) 1 else 0
        )
    }

    constructor(elements: CharIterable) : this(elements.iterator())

    constructor(elements: CharIterator) : this(16.toChar()) {
        while (elements.hasNext()) {
            add(elements.nextChar())
        }
    }

    override fun iterator(): CharIterator {
        return object : OnNextComputeCharIterator() {
            val it = table.asList().listIterator()

            override fun computeNext(): Char {
                var b: Byte

                while (it.hasNext()) {
                    if ((it.next().also { b = it }).toInt() != 0) {
                        return ((it.nextIndex() shl 3) or MathUtil.highestOneBit(b).toInt()).toChar()
                    }
                }

                finish()
                return 0.toChar()
            }
        }
    }

    override val size: Int
        get() = table.asList().parallelStream().mapToInt { i: Byte ->
            Integer.bitCount(i.toInt())
        }.sum()

    override fun add(k: Char): Boolean {
        val index = k.code ushr 3
        val mod = k.code and 7
        table = ByteArrays.ensureCapacity(table, index + 1)
        val prev = table[k.code].toInt()
        val new = table[index].toInt() or (1 shl mod)
        table[index] = new.toByte()
        return prev != new
    }

    override fun contains(k: Char): Boolean {
        val index = k.code ushr 3

        if (index >= table.size) {
            return false
        }

        return ConditionUtil.hasFlag(table[index].toInt(), k.code and 7)
    }

    override fun remove(k: Char): Boolean {
        val index = k.code ushr 3

        if (index >= table.size) {
            return false
        }

        return table[index] != ((1 shl (k.code and 7)).inv().let { table[index] =
            (table[index].toInt() and it).toByte(); table[index] })
    }

    override fun removeAll(c: CharCollection): Boolean {
        var modified = false

        for (s in c) {
            modified = modified or remove(s)
        }

        return modified
    }

    override fun retainAll(c: CharCollection): Boolean {
        return removeIf(CharPredicate { e: Char -> !c.contains(e) })
    }
}
